File: 1646-kth-missing-positive-number/kth-missing-positive-number.ts
function findKthPositive(arr: number[], k: number): number {
    let l = 0;
    let r = arr.length - 1;
    const getNoOfMissing = (i: number): number => (arr[i] - (i + 1));
    while (l <= r) {
        let m = Math.floor((l + r) / 2);
        const missing = getNoOfMissing(m);
        if (missing < k) l = m + 1;
        else r = m - 1;
    }
    if (r === -1) return k;
    return (arr[r] + k - getNoOfMissing(r));

};

File: 1646-kth-missing-positive-number/kth-missing-positive-number.java
class Solution {
    public int findKthPositive(int[] arr, int k) {
        int l = 0;
        int r = arr.length - 1;
        while(l <= r) {
            int m = l + (r - l) / 2;
            int missing = getNumberOfMissing(arr , m);
            if(missing < k) l = m + 1;
            else {
                r = m - 1;
            }
        }
        if(r == -1) return k;
        else return arr[r] + k - getNumberOfMissing(arr , r);
    }
    private int getNumberOfMissing(int[] arr , int i ) {
        return arr[i] - i - 1;
    }
}